### ISSUE TITLE
Infinite Loops possible (and encountered) in fminlbfgs.m

### ISSUE THREAD

**Issue by richardbayes** [gpstuff-dev/gpstuff#14]
I have a method which uses the fminlbfgs.m file frequently.  Occasionally, I have found that infinite loops occur.  I believe these are coming from several lines such as 452, 463, 624, 636, 745, 758 where there are no set maximum iterations.  I believe the infinite loops occur because there are no finite numbers when the loop is started (e.g. all the values are infinite or NaN), and I'm guessing that the functions which take these values are unable to get finite values from these bad starting values.

I've modified the code for my purposes to break the loop after a certain number of iterations and provide a specific error code from which I can make adjustments for my own purposes, but a more general solution to make the code more infinite-loop proof may be useful.

**Comment by avehtari** [comment]
Thanks for reporting. I have previously made several fixes, so that in case of bad values the optimized functions should return NaN's and these should be recognized by fminlbfgs, but I guess you found another case not fixed yet. To find this one, can you be more specific:
 - do you get this with Inf's, NaN's or both?
 - which function you are optimizing, ie, which function is returning Inf's or NaN's?

**Comment by richardbayes** [comment]
Sorry for the delay.  Also, I haven't been able to replicate the error -- I made some changes to my code in the meantime and the issue hasn't been cropping up (I changed the seed that I was using).  I will let you know if it happens again.

I'm using the gp_optim function using the @fminlbfgs handle as the argument to 'optimf'.  I'm obtaining the marginal likelihood of data.   I am also using the gpla_e function, but I believe the error was coming from gp_optim.

Again, sorry for posting an error without being able to replicate.  I replicated it several times earlier, but I made a change somewhere in the meantime.  I'll keep you posted. Many thanks for the excellent code!
